index.rst

```diff
@@ -13,17 +13,256 @@
+This suggests that we might want to make the public interfaces for transfer-from-graph and transfer-from-other-butler more distinct, and possibly make the former private.
+
+``Butler.get``
+""""""""""""""
+
+No journaling is needed, as this is a read-only operation.
+
+``Butler.unstore`` and ``Butler.purge``
+"""""""""""""""""""""""""""""""""""""""
+
+The journal file should be written before ``Registry`` transaction is committed and deleted only after all Datastore deletions succeed.
+This is slightly problematic for client/server, because the journal file will need to be populated with information we get from the Registry database; this means the client cannot be responsible for creating the journal file unless we make fetching the datastore records and deleting them separate operations.
+That isn't too bad - it's just a slight increase in latency and a bit more http traffic.
+Another alternative would be to have the server take responsibility for creating the journal file, and then either returning responsibility for its deletion to the client or taking responsibility for both the deletion of the Datastore artifacts and the deletion of the journal file.
+Which of these is preferable probably depends on whether we want these operations to block until completion and whether we have reasons to perform other Datastore operations on the server (up to this point, having the client use a vanilla ``FileDatastore`` and perform all Datastore operations still seems viable).
+
+.. note::
+   TODO: think about how all of this interacts with DECam raws.
+
+.. note::
+   TODO: think about how all of this affects disassembled composites.
+   Or get rid of them!  I am pretty confident now that we'll never need them, and dropping that before we embark on other major changes to Datastore seems wise (sorry, Tim).
+
+Public interface changes
+========================
+
+.. note::
+   It's all stubs and outlines from here on.
+
+Bundling DatasetRef with Datastore records
+------------------------------------------
+
+- If we're passing around DatasetRefs and bundles of datastore records together a lot, we should have a class for that.
+- This would logically be what a QuantumGraph stores.
+- This + a datastore would be what backs a DeferredDatasetHandle, if we don't want that hitting a butler server or Registry unnecessarily.
+- This might replace FileLocation.
+
+Butler methods vs. Butler.registry methods
+------------------------------------------
+
+- I've totally come around to the idea of Butler being the only provider of public interfaces, and we should move all useful Registry methods there, at the same time we define a richer set of ABCs for various aspects of Butler interfaces.
+- Registry would become a nonpolymorphic concrete class used by one Butler subclass and the Butler server.
```

Comment:
Are you sure? I was coming round to having the public API be `Butler.queryDatasets` (or in fact `Butler.query_datasets` if we are changing things anyhow) but internally that being written as `Butler._registry.query_datasets`. I don't really follow how we end up with a client/server registry at the Butler class level. I thought the purpose of the reorganization was to allow the transactions to be removed from Butler and have atomic Registry APIs that can be switched depending on direct database vs client/server.

Reply:
I am by no means sure, especially after the discussion last Thursday.  The first thing I'm planning to prototype out for this section is now more along the lines of treating Registry as a private polymorphic implementation object, so that we'd have one concrete full Butler class with the full public interface, and different private Registry classes for SQL and client/server.

Reply:
The Registry implementation class is already completely configurable from YAML. You can see that in the client/server registry implementation vs sql registry.

Reply:
My time as an advocate of polymorphic butler was very brief.  Technote is fully back to polymorphic registry as the divide between SQL and client/servert.